[WIP] Experimental feature: Add noise to canvas APIs

elucherini opened the pull request:
#### 192156b7a0b3d25f210fc8b922a7a000b4e69ba8
<pre>
[WIP] Experimental feature: Add noise to canvas APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=243555">https://bugs.webkit.org/show_bug.cgi?id=243555</a>

Reviewed by NOBODY (OOPS!).

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::IdentifierProvider::provider):
(WebCore::IdentifierProvider::identifier const):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
</pre>













































<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a4deeb88f1201a1f5ef989b5f967d7e6668c845

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87274 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31361 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18092 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96389 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149849 "Found 1 new test failure: http/wpt/2dcontext/imagebitmap/noisy-apis.html") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91249 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29726 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25942 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79386 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91277 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92890 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24019 "Found 30 new test failures: editing/pasteboard/paste-double-nested-blockquote-before-blockquote.html, fast/canvas/canvas-ImageData-behaviour.html, fast/canvas/canvas-alphaImageData-behavior.html, fast/canvas/canvas-blend-image.html, fast/canvas/canvas-blend-solid.html, fast/canvas/canvas-blending-global-alpha.html, fast/canvas/canvas-blending-gradient-over-color.html, fast/canvas/canvas-blending-gradient-over-gradient.html, fast/canvas/canvas-blending-pattern-over-color.html, fast/canvas/canvas-blending-pattern-over-pattern.html ...") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74072 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23873 "Found 39 new test failures: fast/canvas/canvas-composite-stroke-alpha.html, fast/canvas/canvas-fillText-invalid-maxWidth.html, fast/canvas/canvas-getImageData.html, fast/canvas/canvas-strokeText-invalid-maxWidth.html, fast/canvas/getPutImageDataPairTest.html, fast/canvas/webgl/gl-pixelstorei.html, fast/canvas/webgl/gl-teximage-imagebitmap-memory.html, fast/canvas/webgl/texImage2DImageDataTest.html, fast/mediastream/video-rotation.html, imported/w3c/web-platform-tests/html/canvas/element/drawing-text-to-the-canvas/2d.text.draw.fill.maxWidth.NaN.html ...") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78993 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79223 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66888 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79131 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27444 "Built successfully") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13059 "Found 30 new test failures: fast/canvas/2d.fillText.gradient.html, fast/canvas/canvas-ImageData-behaviour.html, fast/canvas/canvas-arc-connecting-line.html, fast/canvas/canvas-blend-image.html, fast/canvas/canvas-blend-solid.html, fast/canvas/canvas-blending-clipping.html, fast/canvas/canvas-blending-gradient-over-color.html, fast/canvas/canvas-blending-gradient-over-image.html, fast/canvas/canvas-blending-image-over-image.html, fast/canvas/canvas-blending-image-over-pattern.html ...") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72808 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27395 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14074 "Found 30 new test failures: compositing/iframes/border-uneven-radius-composited-frame.html, fast/canvas/2d.text.draw.fill.maxWidth.verySmall.html, fast/canvas/canvas-alphaImageData-behavior.html, fast/canvas/canvas-blend-solid.html, fast/canvas/canvas-blending-color-over-gradient.html, fast/canvas/canvas-blending-color-over-pattern.html, fast/canvas/canvas-blending-global-alpha.html, fast/canvas/canvas-blending-gradient-over-color.html, fast/canvas/canvas-blending-gradient-over-image.html, fast/canvas/canvas-blending-image-over-color.html ...") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25936 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29080 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36935 "Found 30 new test failures: fast/canvas/canvas-alphaImageData-behavior.html, fast/canvas/canvas-arc-360-winding.html, fast/canvas/canvas-blend-image.html, fast/canvas/canvas-blend-solid.html, fast/canvas/canvas-blending-clipping.html, fast/canvas/canvas-blending-color-over-image.html, fast/canvas/canvas-blending-fill-style.html, fast/canvas/canvas-blending-gradient-over-gradient.html, fast/canvas/canvas-blending-image-over-gradient.html, fast/canvas/canvas-blending-image-over-image.html ...") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75627 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29014 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33351 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16751 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

youssefsoli commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -1016,6 +1016,19 @@ ModelElementEnabled:
       default: false
     WebCore:
       default: false
+      
+NoisyAPIsEnabled:
+  type: bool
+  humanReadableName: "Noisy APIs"
+  humanReadableDescription: "Enable noisy APIs"
+  condition: ENABLE(NOISY_APIS)
+  defaultValue:
+    WebKitLegacy:
+      default: false
+    WebKit:
+      default: false
+    WebCore:
+      default: false
```
I don't think this feature flag is ever used? Currently this feature is set to always build with `#define ENABLE_NOISY_APIS 1`. Since this preference is never checked, the noisy APIs feature will always run, ignoring this flag.

I recommend checking the `Page::settings()` object and how that is used across the code-base, then you can check it inside `CanvasRenderingContext2DBase.cpp`.

elucherini replied:
Should be fixed. (fyi, I updated but it's still a WIP).

youssefsoli replied:
I strongly recommend against the use of `DeprecatedGlobalSettings` if possible, in this case you should be able to have a reference to the page (which comes with a settings object that you can reference instead).

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2180,6 +2185,28 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::createImageData(int sw
+public:
+    static IdentifierProvider& provider()
+    {
+        static NeverDestroyed<FingerprintingProtectionIdentifierProvider> provider;
```
`FingerprintingProtectionIdentifierProvider` isn't defined anywhere

elucherini replied:
Should be fixed. (fyi, I updated but it's still a WIP).

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2222,7 +2249,36 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
+            else if (data.item(i + channel) + fivePercentOffset < 0)
+                data.set(i + channel, 0);
+            else
+                data.set(i + channel, (data.item(i + channel) + fivePercentOffset));
```
We can store `i + channel` and `(data.item(i + channel) + fivePercentOffset)` in a variable for readability and efficiency.

pxlcoder replied:
This can also be rewritten to use `std::clamp`.

elucherini replied:
Stored `i + channel` for now. Will look into `std::clamp`, thanks.

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2222,7 +2249,36 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
+        if (randomNumber() > 0.75) {
+            int fivePercentOffset = static_cast<int>(randomNumber() * 26 - 13);
+            // WTFLogAlways("Random offset: %d", five_percent_offset);
+            size_t channel = static_cast<size_t>(randomNumber() * 3);
```
Do you intend to potentially select the alpha channel? If so, this should be `auto channel = static_cast<size_t>(randomNumber() * 4);`

elucherini replied:
The alpha channel is excluded on purpose.

youssefsoli commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -1341,6 +1341,8 @@
 		44A28AAC12DFB8AC00AE923B /* MathMLElementFactory.h in Headers */ = {isa = PBXBuildFile; fileRef = 44A28AAB12DFB8AC00AE923B /* MathMLElementFactory.h */; };
 		44A28AAF12DFB8BF00AE923B /* MathMLNames.h in Headers */ = {isa = PBXBuildFile; fileRef = 44A28AAE12DFB8BF00AE923B /* MathMLNames.h */; };
 		44AE045C2579CFFB00C42E65 /* AppHighlightRangeData.h in Headers */ = {isa = PBXBuildFile; fileRef = 44AE045A2579CFFB00C42E65 /* AppHighlightRangeData.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		44BC1B2628A1DAE300C7DF2D /* Header.h in Headers */ = {isa = PBXBuildFile; fileRef = 44BC1B2528A1DAE300C7DF2D /* Header.h */; };
```
This `Header.h` entry shouldn't be here.

youssefsoli commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -614,6 +621,15 @@ Document::Document(Frame* frame, const Settings& settings, const URL& url, Docum
+        
+        unsigned hashedDomain = domain.hash();
+        if (hashedDomain)
+            m_identifierProvider->setDomainIdentifier(hashedDomain);
```
`m_identifierProvider` can be null here.

youssefsoli commented on Source/WebCore/page/IdentifierProvider.cpp:
```diff
@@ -0,0 +1,42 @@
+// FIXME: Domain hash must be salted.
+void IdentifierProvider::setDomainIdentifier(const unsigned hashedDomain) {
+    m_domainIdentifier = hashedDomain;
+}
```
Function braces should be in their own line
```suggestion
void IdentifierProvider::setCanvasIdentifier()
{
    // FIXME: Take canvas content, hash it, and combine it with domainIdentifier to create canvasIdentifier
    WTFLogAlways("");
}

// FIXME: Domain hash must be salted.
void IdentifierProvider::setDomainIdentifier(const unsigned hashedDomain)
{
    m_domainIdentifier = hashedDomain;
}
```

youssefsoli commented on Source/WebCore/page/IdentifierProvider.h:
```diff
@@ -0,0 +1,59 @@
+/*
```
This file and the source file should be within `#if ENABLE(NOISY_APIS)`

achristensen07 replied:
I think I disagree.  It would probably be ok to remove all the #if ENABLE(NOISY_APIS) except the one in CanvasRenderingContext2DBase::getImageData.  Adding a class that isn't used and a pointer to the size of Document even if the feature isn't used isn't a high cost for having cleaner code.

youssefsoli commented on Source/WebCore/workers/WorkerThread.h:
```diff
@@ -116,6 +117,8 @@ class WorkerThread : public WorkerOrWorkletThread {
 
     IDBClient::IDBConnectionProxy* idbConnectionProxy();
     SocketProvider* socketProvider();
+    
+    IdentifierProvider* identifierProvider();
```
`identifierProvider` is inconsistently behind the compile flag, I would make sure all instances are covered or else you will run into weird build scenarios.

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2222,7 +2227,38 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
+        IdentifierProvider* identifierProvider = canvasBase().scriptExecutionContext()->identifierProvider();
+        // Using WeakRandom as I could explicitly set the seed.
+        // FIXME: set and use canvasIdentifier instead of domainIdentifier.
+        WeakRandom randomNumberGenerator = WeakRandom(identifierProvider->domainIdentifier());
```
`auto`

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -76,6 +77,10 @@
 
+#if ENABLE(NOISY_APIS)
+#include <wtf/WeakRandom.h>
+#endif // ENABLE(NOISY_APIS)
```
You don't need the comment if the build flag spans a couple of lines.

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2222,7 +2227,38 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
+        IdentifierProvider* identifierProvider = canvasBase().scriptExecutionContext()->identifierProvider();
+        // Using WeakRandom as I could explicitly set the seed.
+        // FIXME: set and use canvasIdentifier instead of domainIdentifier.
+        WeakRandom randomNumberGenerator = WeakRandom(identifierProvider->domainIdentifier());
```
Why aren't we using the `canvasIdentifier` here? (or rather, what is the purpose of the `canvasIdentifier`)

elucherini replied:
As per Slack conversation

youssefsoli commented on Source/WebCore/page/IdentifierProvider.cpp:
```diff
@@ -0,0 +1,42 @@
+
+#include "IdentifierProvider.h"
+
+//#include "RegistrableDomain.h"
```
Try to remove all the comments before final review.

youssefsoli commented on Source/WebCore/page/IdentifierProvider.cpp:
```diff
@@ -0,0 +1,42 @@
+
+void IdentifierProvider::setCanvasIdentifier() {
+    // FIXME: Take canvas content, hash it, and combine it with domainIdentifier to create canvasIdentifier
+    WTFLogAlways("");
```
Same thing with all of the `WTFLog` statements, remove them before final review. If you want logging, search the codebase for `LOG(` or take a look at classes that inherit `LoggerHelper`.

youssefsoli commented on Source/WebCore/page/IdentifierProvider.h:
```diff
@@ -0,0 +1,59 @@
+    : m_domainIdentifier { 0 }
+    , m_canvasIdentifier { 0 }
+    {
+    }
```
This can just be 
```suggestion
    IdentifierProvider() = default;
```
and setting the initial values in the class like
```
private:
     unsigned m_domainIdentifier { 0 };
```

youssefsoli commented on Source/WebCore/page/IdentifierProvider.h:
```diff
@@ -0,0 +1,59 @@
+//#include "RegistrableDomain.h"
+#include <wtf/ThreadSafeRefCounted.h>
+#include <wtf/URL.h>
+#include <wtf/WeakRandom.h>
```
Re-evaluate what headers are required and what can be forward declared.

youssefsoli commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2222,7 +2227,38 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
+        IdentifierProvider* identifierProvider = canvasBase().scriptExecutionContext()->identifierProvider();
+        // Using WeakRandom as I could explicitly set the seed.
+        // FIXME: set and use canvasIdentifier instead of domainIdentifier.
+        WeakRandom randomNumberGenerator = WeakRandom(identifierProvider->domainIdentifier());
```
`identifierProvider` can be null here.

elucherini replied:
Which cases will this be null?

youssefsoli reviewed (changes requested):
Is the goal of the `IdentifierProvider` to be a source of consistent randomness for APIs that choose to be noisy?
If so, it might make more sense for the random number generation to happen within that class as an easy to use API (maybe no longer naming it `IdentifierProvider`) rather than having each client instantiate their own random number generator.

Which leads to a similar question, are these identifiers planning to be used in anything other than predictable randomness?

achristensen07 commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -9340,6 +9342,9 @@
 		44AE045A2579CFFB00C42E65 /* AppHighlightRangeData.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = AppHighlightRangeData.h; sourceTree = "<group>"; };
 		44B38BF42536901A00A4458D /* SelectionRangeData.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = SelectionRangeData.h; sourceTree = "<group>"; };
 		44B38BF925369A8800A4458D /* SelectionRangeData.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = SelectionRangeData.cpp; sourceTree = "<group>"; };
+		44BC1B2528A1DAE300C7DF2D /* Header.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = Header.h; path = ../../../Header.h; sourceTree = "<group>"; };
```
This seems suspicious.  Could you remove all references to Header.h?  I don't think it's actually added.

achristensen07 commented on Source/WebCore/dom/Document.h:
```diff
@@ -734,6 +734,10 @@ class Document
     RefPtr<RTCDataChannelRemoteHandlerConnection> createRTCDataChannelRemoteHandlerConnection() final;
 
+#if ENABLE(NOISY_APIS)
+        IdentifierProvider* identifierProvider() final;
```
Too much indentation.

achristensen07 commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2222,7 +2227,38 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
+            if (randomNumberGenerator.get() > 0.75) {
+                int fivePercentOffset = static_cast<int>(randomNumberGenerator.get() * 26 - 13);
+                size_t channel = i + static_cast<size_t>(randomNumberGenerator.get() * 3);
+                // WTFLogAlways("Random channel: %zu", channel);
```
I know this is WIP, but please don't add any WTFLogAlways calls when the final PR is up.

achristensen07 commented on Source/WebCore/page/IdentifierProvider.cpp:
```diff
@@ -0,0 +1,42 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "IdentifierProvider.h"
```
This file is currently unnecessary.

achristensen07 commented on Source/WebCore/page/IdentifierProvider.cpp:
```diff
@@ -0,0 +1,42 @@
+
+namespace WebCore {
+
+void IdentifierProvider::setCanvasIdentifier() {
```
Use WebKit style and have the { on the next line.

achristensen07 commented on Source/WebCore/page/IdentifierProvider.h:
```diff
@@ -0,0 +1,59 @@
+    { return adoptRef(*new IdentifierProvider()); }
+
+    IdentifierProvider()
+    : m_domainIdentifier { 0 }
```
The indentation here is inconsistent with the rest of WebKit, and I think this constructor is unnecessary.  It can be replaced by initializer lists lower like "unsigned m_domainIdentifier { 0 };"

achristensen07 commented on Source/WebCore/page/PageConfiguration.h:
```diff
@@ -83,7 +83,8 @@ class WebRTCProvider;
     WTF_MAKE_NONCOPYABLE(PageConfiguration); WTF_MAKE_FAST_ALLOCATED;
 public:
-    WEBCORE_EXPORT PageConfiguration(PAL::SessionID, UniqueRef<EditorClient>&&, Ref<SocketProvider>&&, UniqueRef<WebRTCProvider>&&, Ref<CacheStorageProvider>&&, Ref<UserContentProvider>&&, Ref<BackForwardClient>&&, Ref<CookieJar>&&, UniqueRef<ProgressTrackerClient>&&, UniqueRef<FrameLoaderClient>&&, UniqueRef<SpeechRecognitionProvider>&&, UniqueRef<MediaRecorderProvider>&&, Ref<BroadcastChannelRe ...
+    WEBCORE_EXPORT PageConfiguration(PAL::SessionID, UniqueRef<EditorClient>&&, Ref<SocketProvider>&&, UniqueRef<WebRTCProvider>&&, Ref<CacheStorageProvider>&&, Ref<UserContentProvider>&&, Ref<BackForwardClient>&&, Ref<CookieJar>&&, UniqueRef<ProgressTrackerClient>&&, UniqueRef<FrameLoaderClient>&&, UniqueRef<SpeechRecognitionProvider>&&, UniqueRef<MediaRecorderProvider>&&, Ref<BroadcastChannelRe ...
```
I don't think you need to touch PageConfiguration.h at all.

achristensen07 commented on Source/WebCore/workers/WorkerMessagingProxy.cpp:
```diff
@@ -124,14 +124,20 @@ void WorkerMessagingProxy::startWorkerGlobalScope(const URL& scriptURL, PAL::Ses
     SocketProvider* socketProvider = document.socketProvider();
+    
+#if ENABLE(NOISY_APIS)
+    IdentifierProvider* identifierProvider = document.identifierProvider();
```
You could just have document.identifierProvider return null if ENABLE(NOISY_APIS) is 0 instead of having a check here.

youssefsoli commented on Source/WebCore/page/IdentifierProvider.cpp:
```diff
@@ -0,0 +1,51 @@
+    // FIXME: Use computeHash() template instead.
+    StringHasher hasher;
+    
+    for (size_t i = 0; i < canvasData.length(); i += 4)
+        hasher.addCharacter((UChar)canvasData.item(i));
```
Why are we only hashing with the first channel? This makes a hash collision trivially easy.

A simple attack would be to swap the red and alpha channels (255 all across), read it out. Then have another canvas with the same red channel (255 all across) hash that you know the pixel values of and you can reverse the noise.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -614,6 +621,21 @@ Document::Document(Frame* frame, const Settings& settings, const URL& url, Docum
     setStorageBlockingPolicy(m_settings->storageBlockingPolicy());
+        
+#if ENABLE(NOISY_APIS)
+    m_identifierProvider->setShouldUseNoisyApis(page()->settings().noisyAPIsEnabled());
```
page needs a null check like it had above.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -614,6 +621,21 @@ Document::Document(Frame* frame, const Settings& settings, const URL& url, Docum
+#if ENABLE(NOISY_APIS)
+    m_identifierProvider->setShouldUseNoisyApis(page()->settings().noisyAPIsEnabled());
+    // FIXME: Only call for mainframe.
+    if (isHTMLDocument() && !m_url.isEmpty()) {
```
If settings().noisyAPIsEnabled() returns false, we don't want to do what is inside this if statement.

achristensen07 commented on LayoutTests/http/wpt/2dcontext/imagebitmap/noisy-apis.html:
```diff
@@ -0,0 +1,51 @@
+            initializeSolidCanvas(canvas, 255,0,0,255);
+            resolve(canvas);
+        }).then(function(canvas) {
+            _assert(testCanvasPixelsDifferent(canvas, 255,0,0,255), "testCanvasPixelsDifferent");
```
This will fail about 2^-32 of the time.  That may be something we don't care about, but adding a function to Internals.idl that calls setDomainIdentifier with a known deterministic value would prevent that and also allow us to measure the exact values of the pixels instead of just testing that they are close to but not equal to this value.

achristensen07 commented on Source/WebCore/page/IdentifierProvider.h:
```diff
@@ -0,0 +1,56 @@
+private:
+    uint64_t m_domainIdentifier { 0 };
+    uint64_t m_canvasIdentifier { 0 };
+    bool m_shouldUseNoisyApis { false };
```
nit: m_shouldUseNoisyAPIs

Commit:
Experimental feature: Add noise to canvas APIs
https://bugs.webkit.org/show_bug.cgi?id=243555

Reviewed by NOBODY (OOPS!).

* LayoutTests/http/wpt/2dcontext/imagebitmap/common.js:
(testCanvasPixelsDifferent):
(initializeSolidCanvas):
* LayoutTests/http/wpt/2dcontext/imagebitmap/noisy-apis-expected.txt: Added.
* LayoutTests/http/wpt/2dcontext/imagebitmap/noisy-apis.html: Added.
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::m_whitespaceCache):
(WebCore::Document::identifierProvider):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getImageData const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/page/IdentifierProvider.cpp: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
(WebCore::IdentifierProvider::setCanvasIdentifier):
(WebCore::IdentifierProvider::setDomainIdentifier):
* Source/WebCore/page/IdentifierProvider.h: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebCore/page/Page.cpp:
(WebCore::m_identifierProvider):
(WebCore::m_contentSecurityPolicyModeForExtension): Deleted.
* Source/WebCore/page/Page.h:
(WebCore::Page::identifierProvider):
* Source/WebCore/page/PageConfiguration.cpp:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::create):
(WebCore::DedicatedWorkerGlobalScope::DedicatedWorkerGlobalScope):
* Source/WebCore/workers/DedicatedWorkerGlobalScope.h:
* Source/WebCore/workers/DedicatedWorkerThread.cpp:
(WebCore::DedicatedWorkerThread::DedicatedWorkerThread):
(WebCore::DedicatedWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/DedicatedWorkerThread.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::WorkerGlobalScope):
(WebCore::WorkerGlobalScope::identifierProvider):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WebCore::WorkerMessagingProxy::startWorkerGlobalScope):
* Source/WebCore/workers/WorkerThread.cpp:
(WebCore::WorkerThread::WorkerThread):
(WebCore::WorkerThread::identifierProvider):
* Source/WebCore/workers/WorkerThread.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::create):
(WebCore::ServiceWorkerGlobalScope::ServiceWorkerGlobalScope):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::ServiceWorkerThread):
(WebCore::ServiceWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
* Source/WebCore/workers/shared/SharedWorkerGlobalScope.cpp:
(WebCore::SharedWorkerGlobalScope::SharedWorkerGlobalScope):
* Source/WebCore/workers/shared/SharedWorkerGlobalScope.h:
* Source/WebCore/workers/shared/context/SharedWorkerThread.cpp:
(WebCore::SharedWorkerThread::SharedWorkerThread):
(WebCore::SharedWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/shared/context/SharedWorkerThread.h:
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.cpp:
(WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy):
* Source/WebCore/worklets/WorkletGlobalScope.h:
